python/arb.cpp

```diff
@@ -0,0 +1,235 @@
+#include <vector>
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include <domain_decomposition.hpp>
+#include <model.hpp>
+#include <load_balance.hpp>
+#include <profiling/meter_manager.hpp>
+#include <recipe.hpp>
+#include <rss_cell.hpp>
+
+#include "print.hpp"
+#include "recipe.hpp"
+
+namespace pb = pybind11;
+
+template <typename T>
+arb::util::any wrap_any(T value) {
+    return arb::util::any(std::move(value));
+}
+
+struct spike_recorder {
+    using f = arb::model::spike_export_function;
+    using spike_vec = std::vector<arb::spike>;
+    spike_vec spike_store;
+
+    f callback() {
+        return [&](const spike_vec& spikes) {
+            spike_store.insert(spike_store.end(), spikes.begin(), spikes.end());
+        };
+    }
+};
+
+std::unique_ptr<spike_recorder> make_spike_recorder(arb::model& m) {
+    auto r = arb::util::make_unique<spike_recorder>();
+    m.set_global_spike_callback(r->callback());
+    return r;
+}
+
+// helpful string literals that reduce verbosity
+using namespace pybind11::literals;
+
+PYBIND11_MODULE(pyarb, m) {
+    //
+    // common types
+    //
+
+    pb::class_<arb::cell_member_type> cell_member(m, "cell_member",
+        "For global identification of an item of cell local data.\n\n"
+        "Items of cell_member_type must:\n"
+        "(1) be associated with a unique cell, identified by the member gid;\n"
+        "(2) identify an item within a cell-local collection by the member index.\n");
+
+    cell_member
+        .def(pb::init<>())
+        .def_readwrite("index", &arb::cell_member_type::index,
+            "Cell-local index of the item.")
+        .def_readwrite("gid",   &arb::cell_member_type::gid,
+            "The global identifier of a cell.")
+        .def("__str__",  &cell_member_string)
+        .def("__repr__", &cell_member_string);
+
+    // spike recording
+    pb::class_<spike_recorder> spike_recorder(m, "spike_recorder");
+    spike_recorder
+        .def(pb::init<>())
+        .def_readonly("spikes", &spike_recorder::spike_store);
+    m.def("make_spike_recorder", &make_spike_recorder);
+
```

Comment:
I think there is the risk of a spike recorder being GCed and the corresponding implicit reference to the spike vector going away while the model is running.

For python purposes, I think it's worth considering having a stack or list of spike sink callbacks, with the corresponding objects deregistering themselves in the event of their destruction (or have reference counting with e.g. `shared_ptr`.) The infrastructure for this could be python-library specific, or we could add it as a feature in the Arbor library that sits on top of the model `set_global_spike_callback` interface. Callbacks could even be python functions.

There is also the question of global vs local spikes for this sort of interface.

Reply:
Well, if the user somehow conspires to have their spike recorder get GC'ed during model execution, they won't be able to use the output.

Reply:
True, but we should strive to avoid memory corruption or segfaults regardless :)